force-app/main/default/classes/CountTasksTest.cls

```diff
@@ -0,0 +1,183 @@
+      [SELECT COUNT() FROM Task],
+      (ACCOUNTS_NUMBER * TASKS_NUMBER * numberOfTaskTypes),
+      'Failed to get expected amount of tasks'
+    );
+  }
+
+  //insert scenario
+  @isTest
+  public static void insertNewTaskAccount() {
+    Account account = [SELECT Id FROM Account LIMIT 1];
+    Id accId = account.Id;
+
+    Task tskOther = new Task(Subject = 'Other', Type = 'Other', WhatId = accId);
+    insert tskOther;
+
+    Assert.areEqual(
+      [SELECT Open_Tasks_Other__c FROM Account WHERE Id = :accId]
+      .Open_Tasks_Other__c,
+      TASKS_NUMBER + 1,
+      'Failed to get expected amount of open tasks on Account after insert'
+    );
+  }
+
+  //close a task & re-open it scenario
+  @isTest
+  private static void closeTaskAndReopen() {
+    //closing
+    Account account = [SELECT Id FROM Account LIMIT 1];
+    Id accId = account.Id;
+
+    Task openTsk = [
+      SELECT Type, Status
+      FROM Task
+      WHERE WhatId = :accId AND Status != 'Completed' AND Type = 'Email'
+      LIMIT 1
+    ];
+    openTsk.Status = 'Completed';
+    update openTsk;
+
+    Assert.areEqual(
+      [SELECT Open_Tasks_Email__c FROM Account WHERE Id = :accId]
+      .Open_Tasks_Email__c,
+      TASKS_NUMBER - 1,
+      'Failed to get expected amount of open tasks on Account after closing a task'
+    );
+
+    //re-opening
+    openTsk.Status = 'In progress';
+    update openTsk;
+
+    Assert.areEqual(
+      [SELECT Open_Tasks_Email__c FROM Account WHERE Id = :accId]
+      .Open_Tasks_Email__c,
+      TASKS_NUMBER,
+      'Failed to get expected amount of open tasks on Account after re-opening a task'
+    );
+  }
+
+  //delete scenario
+  @isTest
+  private static void deleteTask() {
+    Account account = [SELECT Id FROM Account LIMIT 1];
+    Id accId = account.Id;
+
+    Task openTsk = [
+      SELECT Type, Status
+      FROM Task
+      WHERE WhatId = :accId AND Status != 'Completed' AND Type = 'Other'
+      LIMIT 1
+    ];
+
+    Assert.areEqual(
+      [SELECT Open_Tasks_Other__c FROM Account WHERE Id = :accId]
+      .Open_Tasks_Other__c,
+      TASKS_NUMBER,
+      'Failed to get expected amount of open tasks on Account before delete'
+    );
+
+    delete openTsk;
+
+    Assert.areEqual(
+      [SELECT Open_Tasks_Other__c FROM Account WHERE Id = :accId]
+      .Open_Tasks_Other__c,
+      TASKS_NUMBER - 1,
+      'Failed to get expected amount of open tasks on Account after delete'
+    );
+  }
+
+  //reparenting a task to a new account scenario (wrong count)
+  @isTest
+  private static void reparentingTask() {
+    Account account = [SELECT Id FROM Account LIMIT 1];
+    Id accId = account.Id;
+
+    Task openTsk = [
+      SELECT Type, Status
+      FROM Task
+      WHERE WhatId != :accId AND Status != 'Completed' AND Type = 'Other'
+      LIMIT 1
+    ];
+    Id firstParent = openTsk.WhatId;
+    openTsk.WhatId = accId;
+
+    update openTsk;
+
+    //the count is done wrong on the first related Account
+    Assert.areNotEqual(
```

Comment:
Thanks. As you might have guessed, the next challenge is to analyze why this is not working with the current implementation. No fixes expected yet, explanation is our goal.

Reply:
This is not working with the current implementation because we have an _after_ trigger, so the task which fires the trigger is already updated with the new accountId/whatId and the count will be done only on the new account (based on the business logic I have now).

To fix this, I think I should do a comparison between oldMap and newMap (so between the old version of a task and the new one) and when the accountId of the Task in the newMap is different from the accountId in the oldMap I should also do the count on the account in oldMap.

Reply:
Yes, you are going into the right direction.

There is a simple, straight-forward (I think) solution as well. Would you like to try it or should I reveal my approach?

Reply:
I would like to try it myself first :)

Reply:
I was thinking to do this:
![image](https://user-images.githubusercontent.com/77618420/217549733-34327f6f-f9bb-495a-b7bf-58817a7f3882.png)

I don't know if it's the best solution, because we said that in the Trigger we should have less logic.